### ISSUE TITLE
Question about contact_points.json and penetration_points.json

### ISSUE THREAD

**Issue by Dingry** [PKU-EPIC/DexGraspNet#8]
Hi, thanks for your great work. I'm interested in utilizing AbilityHand to generate some grasping poses. I've observed that there are contact_points.json and penetration_points.json files for the robot hand. Could you kindly explain the meaning of the values in these files? Also, how could I obtain these for a custom robot hand such as Ability Hand? Thank you very much for your time and assistance.

https://github.com/PKU-EPIC/DexGraspNet/blob/main/grasp_generation/mjcf/contact_points.json
https://github.com/PKU-EPIC/DexGraspNet/blob/main/grasp_generation/mjcf/penetration_points.json

**Comment by COMoER** [comment]
I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. 
![image](https://github.com/PKU-EPIC/DexGraspNet/assets/58421330/b01dac1c-f5f6-42c8-8972-5a8febef87f0)

But I found that some of them are even out of hand mesh. It is strange.

**Comment by mzhmxzh** [comment]
> I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. ![image](https://user-images.githubusercontent.com/58421330/280738320-b01dac1c-f5f6-42c8-8972-5a8febef87f0.png)
> 
> But I found that some of them are even out of hand mesh. It is strange.

Which version of trimesh are you using? 
Try switching to a lower version like 3.10.8.

**Comment by Dingry** [comment]
> I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. ![image](https://user-images.githubusercontent.com/58421330/280738320-b01dac1c-f5f6-42c8-8972-5a8febef87f0.png)
> 
> But I found that some of them are even out of hand mesh. It is strange.

Hello, could you kindly explain how you acquired these contact points? Could you share some insights? Thank you very much.

**Comment by COMoER** [comment]
> > I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. ![image](https://user-images.githubusercontent.com/58421330/280738320-b01dac1c-f5f6-42c8-8972-5a8febef87f0.png)
> > But I found that some of them are even out of hand mesh. It is strange.
> 
> Hello, could you kindly explain how you acquired these contact points? Could you share some insights? Thank you very much.

I just visualized the contact points provided by the authors. I modified the visualization function in `grasp_generation/utils/hand_model.py` as:
```
    def get_open3d_data(self, i,  with_contact_points=False, pose=None):
        if pose is not None:
            pose = np.array(pose, dtype=np.float32)
        data = trimesh.Trimesh()
        for link_name in self.mesh:
            v = self.current_status[link_name].transform_points(self.mesh[link_name]['vertices'])
            if len(v.shape) == 3:
                v = v[i]
            v = v @ self.global_rotation[i].T + self.global_translation[i]
            v = v.detach().cpu()
            f = self.mesh[link_name]['faces'].detach().cpu()
            if pose is not None:
                v = v @ pose[:3, :3].T + pose[:3, 3]

            data += trimesh.Trimesh(v,f)

        if with_contact_points:

            contact_points = self.penetration_keypoints

            if pose is not None:
                contact_points = contact_points @ pose[:3, :3].T + pose[:3, 3]
            pcd = o3d.geometry.PointCloud()
            pcd.points = o3d.utility.Vector3dVector(contact_points)
            o3d.io.write_point_cloud("2.ply",pcd)
            # data.append(
            #     go.Scatter3d(x=contact_points[:, 0], y=contact_points[:, 1], z=contact_points[:, 2], mode='markers',
            #                  marker=dict(color='red', size=5)))
        o3d.io.write_triangle_mesh("1.ply",data.as_open3d)
```
And visualize the meshes through meshlab.

**Comment by mzhmxzh** [comment]
> > I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. ![image](https://user-images.githubusercontent.com/58421330/280738320-b01dac1c-f5f6-42c8-8972-5a8febef87f0.png)
> > But I found that some of them are even out of hand mesh. It is strange.
> 
> Hello, could you kindly explain how you acquired these contact points? Could you share some insights? Thank you very much.

* First, manually determine the regions of the robotic hand that are likely to contact the object during grasping. For instance, the front of the proximal link of the forefinger. 
* Second, manually and evenly sample some points in each region. 
* Third, record the contact candidates as a list of points, and store these lists as a dictionary in contact_points.json. All contact candidates are recorded in the link's own reference frame, rather than the world's reference frame. When you need to calculate their positions in the world's reference frame, you can use the forward kinematics that we've implemented in class HandModel. Refer to that class for more details. 

I think the most laborious part is sampling the contact points. Here are some tips to make it faster. The collision mesh of ShadowHand is composed of boxes and capsules. For each capsule, I first determine the starting/terminating height and the starting/terminating angle of the contact region. Then, I determine how many rows and columns of points I want to sample in that region. Finally, I use the radius and height of that capsule, and write a short script to sample the contact points for that link. For instance, if the capsule is upright, and the bottom sphere is located at the origin, then a point with height `z` and angle `theta` is located at `[radius * cos(theta), radius * sin(theta), z]`. 

Also, you can refer to the `allegro` branch of the repo, which gives an example of how to customize the repo to another robotic hand. The collision mesh of Allegro is composed of multiple boxes. So I wrote a script to sample points in a region given the box's parameters and the region of interest. For your reference, it took me roughly 3 hours to develop the Allegro repo from the ShadowHand repo. 

The first thing I suggest doing is visualizing the collision mesh of the AbilityHand. If you only have the visual mesh, I strongly suggest designing a collision primitive for each link. Because our source code has acceleration for calculating the signed distance function of capsules. And because sampling contact points is much easier on primitives. Or else you would need a modeling software like Blender to select contact regions and sample points. What I used to do was, export the mesh of each link in its own reference frame, then open it in Blender and delete the facets outside the contact region, and use pytorch3d to sample points on the mesh of the contact region with furthest point sampling.

**Comment by mzhmxzh** [comment]
> > > I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. ![image](https://user-images.githubusercontent.com/58421330/280738320-b01dac1c-f5f6-42c8-8972-5a8febef87f0.png)
> > > But I found that some of them are even out of hand mesh. It is strange.
> > 
> > 
> > Hello, could you kindly explain how you acquired these contact points? Could you share some insights? Thank you very much.
> 
> I just visualized the contact points provided by the authors. I modified the visualization function in `grasp_generation/utils/hand_model.py` as:
> 
> ```
>     def get_open3d_data(self, i,  with_contact_points=False, pose=None):
>         if pose is not None:
>             pose = np.array(pose, dtype=np.float32)
>         data = trimesh.Trimesh()
>         for link_name in self.mesh:
>             v = self.current_status[link_name].transform_points(self.mesh[link_name]['vertices'])
>             if len(v.shape) == 3:
>                 v = v[i]
>             v = v @ self.global_rotation[i].T + self.global_translation[i]
>             v = v.detach().cpu()
>             f = self.mesh[link_name]['faces'].detach().cpu()
>             if pose is not None:
>                 v = v @ pose[:3, :3].T + pose[:3, 3]
> 
>             data += trimesh.Trimesh(v,f)
> 
>         if with_contact_points:
> 
>             contact_points = self.penetration_keypoints
> 
>             if pose is not None:
>                 contact_points = contact_points @ pose[:3, :3].T + pose[:3, 3]
>             pcd = o3d.geometry.PointCloud()
>             pcd.points = o3d.utility.Vector3dVector(contact_points)
>             o3d.io.write_point_cloud("2.ply",pcd)
>             # data.append(
>             #     go.Scatter3d(x=contact_points[:, 0], y=contact_points[:, 1], z=contact_points[:, 2], mode='markers',
>             #                  marker=dict(color='red', size=5)))
>         o3d.io.write_triangle_mesh("1.ply",data.as_open3d)
> ```
> 
> And visualize the meshes through meshlab.

Solution: `pip install trimesh==3.10`

First, `self.penetration_keypoints` is recorded in each link's local reference frame. You need forward kinematics to convert them into the world frame. Please refer to self.get_penetraion_keypoints for more details. Also refer to https://github.com/PKU-EPIC/DexGraspNet/blob/main/grasp_generation/tests/visualize_hand_model.py for visualization. 

Second, what I am really concerned is the collision mesh of the ShadowHand. It shouldn't look like this. The problem is actually with trimesh. In lower versions of trimesh, e.g. 3.10, when you create a capsule using trimesh.primitives.Capsule, the origin coincides with the center of the lower sphere. In higher versions of trimesh, e.g. 3.22, the origin coincides with the center of symmetry of the entire capsule, rather than the lower sphere of the capsule. So when the forward kinematics module assembles the link pieces, the result gets weird. I've seen many people who have encountered this issue, and `pip install trimesh==3.10` solved it nicely. If you wish to use the newest version of trimesh, then you would want to translate the capsule on creation, so that the lower sphere lies at the origin.

**Comment by Dingry** [comment]
Thanks so much for your advice. It seems that for the allegro hand, the penetration points are not used during grasp pose generation. I would like to know why it is different from the Shadow Hand.

**Comment by mzhmxzh** [comment]
For the self penetration energy of ShadowHand, we manually selected a couple of penetration key points, and set a self penetration threshold of 2cm. If two key points are closer than 2cm during optimization, then the hand pose is given a self penetration penalty. We made sure that the penetration energy is zero in all valid hand poses. You can check the visualization of the self penetration key points in https://github.com/PKU-EPIC/DexGraspNet/blob/main/grasp_generation/tests/visualize_hand_model.py. 

Although this energy cannot detect all self penetration, our experiments proved it sufficient for detecting and avoiding most self penetration for ShadowHand. However, Allegro has much thicker fingers, and its fingers are boxes rather than capsules. So using key-points-based self penetration energy isn't sufficient. That's why we used signed-distance-function-based self penetration energy for the Allegro hand. 

Specifically, 1000 surface points are sampled on the Allegro hand with furthest point sampling. Then, we calculate the signed distance from each surface point to the Allegro hand. A self collision mask is set to filter the link pairs that shouldn't be accounted in the self penetration energy. Finally, the sum of the penetration distances of all surface points is used as the self penetration energy. This energy is slow to compute, but it can effectively detect and avoid any self penetration. You can check the surface points of the Allegro hand here https://github.com/PKU-EPIC/DexGraspNet/blob/allegro/grasp_generation/tests/visualize_hand_model.py.

**Comment by Dingry** [comment]
Thanks so much for your advice and help!

**Comment by COMoER** [comment]
> > > > I have the same question. Besides, I have tried to visualize the contact points to find the meaning as the following figure. ![image](https://user-images.githubusercontent.com/58421330/280738320-b01dac1c-f5f6-42c8-8972-5a8febef87f0.png)
> > > > But I found that some of them are even out of hand mesh. It is strange.
> > > 
> > > 
> > > Hello, could you kindly explain how you acquired these contact points? Could you share some insights? Thank you very much.
> > 
> > 
> > I just visualized the contact points provided by the authors. I modified the visualization function in `grasp_generation/utils/hand_model.py` as:
> > ```
> >     def get_open3d_data(self, i,  with_contact_points=False, pose=None):
> >         if pose is not None:
> >             pose = np.array(pose, dtype=np.float32)
> >         data = trimesh.Trimesh()
> >         for link_name in self.mesh:
> >             v = self.current_status[link_name].transform_points(self.mesh[link_name]['vertices'])
> >             if len(v.shape) == 3:
> >                 v = v[i]
> >             v = v @ self.global_rotation[i].T + self.global_translation[i]
> >             v = v.detach().cpu()
> >             f = self.mesh[link_name]['faces'].detach().cpu()
> >             if pose is not None:
> >                 v = v @ pose[:3, :3].T + pose[:3, 3]
> > 
> >             data += trimesh.Trimesh(v,f)
> > 
> >         if with_contact_points:
> > 
> >             contact_points = self.penetration_keypoints
> > 
> >             if pose is not None:
> >                 contact_points = contact_points @ pose[:3, :3].T + pose[:3, 3]
> >             pcd = o3d.geometry.PointCloud()
> >             pcd.points = o3d.utility.Vector3dVector(contact_points)
> >             o3d.io.write_point_cloud("2.ply",pcd)
> >             # data.append(
> >             #     go.Scatter3d(x=contact_points[:, 0], y=contact_points[:, 1], z=contact_points[:, 2], mode='markers',
> >             #                  marker=dict(color='red', size=5)))
> >         o3d.io.write_triangle_mesh("1.ply",data.as_open3d)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And visualize the meshes through meshlab.
> 
> Solution: `pip install trimesh==3.10`
> 
> First, `self.penetration_keypoints` is recorded in each link's local reference frame. You need forward kinematics to convert them into the world frame. Please refer to self.get_penetraion_keypoints for more details. Also refer to https://github.com/PKU-EPIC/DexGraspNet/blob/main/grasp_generation/tests/visualize_hand_model.py for visualization.
> 
> Second, what I am really concerned is the collision mesh of the ShadowHand. It shouldn't look like this. The problem is actually with trimesh. In lower versions of trimesh, e.g. 3.10, when you create a capsule using trimesh.primitives.Capsule, the origin coincides with the center of the lower sphere. In higher versions of trimesh, e.g. 3.22, the origin coincides with the center of symmetry of the entire capsule, rather than the lower sphere of the capsule. So when the forward kinematics module assembles the link pieces, the result gets weird. I've seen many people who have encountered this issue, and `pip install trimesh==3.10` solved it nicely. If you wish to use the newest version of trimesh, then you would want to translate the capsule on creation, so that the lower sphere lies at the origin.

Thx! After installing trimesh==3.10, the visualization of contact points is reasonable.
![image](https://github.com/PKU-EPIC/DexGraspNet/assets/58421330/ddb8ab08-41fd-479a-8181-5997c3380f38)

**Comment by kaijieshi7** [comment]
Hi, @Dingry , did you succeed to generate grasp pose for ability hand?

**Comment by Dingry** [comment]
> Hi, @Dingry , did you succeed to generate grasp pose for ability hand?

Yes. I generate contact points myself for the ability hand.
